driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -1170,55 +1154,170 @@ int getNumPinnedToTransaction() {
         }
     }
 
+    /**
+     * This class maintains threads needed to perform {@link #getAsync(SingleResultCallback)}.
+     */
     @ThreadSafe
-    private static class Workers implements AutoCloseable {
-        private volatile ExecutorService getter;
-        private volatile ExecutorService opener;
+    private static class AsyncWorkManager implements AutoCloseable {
+        private volatile State state;
+        private volatile BlockingQueue<Task> tasks;
         private final Lock lock;
+        @Nullable
+        private ExecutorService worker;
 
-        Workers() {
+        AsyncWorkManager() {
+            state = State.NEW;
+            tasks = new LinkedBlockingQueue<>();
             lock = new StampedLock().asWriteLock();
```

Comment:
I realized I don't understand why we're using StampedLock here.  Can you explain?

Reply:
Picking the functionally simplest tool for the task. Alternative would be `ReentrantLock`, which adds the following functionality:

1. fairness
2. reentrancy, which requires maintaining the owner as part of the lock object state
3. conditional variables

Since neither is needed for this use case, `ReentrantLock` is not needed.